src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs

```diff
@@ -0,0 +1,202 @@
+            .Select(_ => new UnspentOutputTable
+            {
+               Address = _.Address, Outpoint = _.Outpoint, Value = _.Value, BlockIndex = -1
+            })
+            .ToList();
+
+         if (!filteredItemsToCopy.Any())
+            break;
+
+         await storage.UnspentOutputTable.InsertManyAsync(filteredItemsToCopy);
+         moreItemsToCopy = itemsToCopy.Count == lookupItems.Count;
+
+
+      } while (moreItemsToCopy);
+   }
+
+   private static async Task DeleteBlockInCollectionFromTopOfTable<T>(IMongoCollection<T> collection,
+      string propertyName, long blockIndex)
+   {
+      const int limit = 10000;
+
+      do
+      {
+         var lookupItems = await GetTopNDocumentsFromCollectionAsync<T, BsonDocument>(collection, 0, limit);
+
+         if (!lookupItems.Any())
+            break;
+
+         foreach (BsonDocument bsonDocument in lookupItems)
+         {
+            if (bsonDocument[propertyName] != blockIndex)
+               return;
+
+            await collection.FindOneAndDeleteAsync(FilterDefinition<T>.Empty,
+               new FindOneAndDeleteOptions<T>
+               {
+                  Sort = new BsonDocumentSortDefinition<T>(new BsonDocument("_id", -1))
+               });
+         }
+
+      } while (true);
+   }
+
+   /// <summary>
+   /// Inputs spend outputs, when an output is spent it gets deleted from the UnspendOutput table and the action of the delete is represented in the inputs table,
+   /// when a rewind happens we need to bring back outputs that have been deleted from the UnspendOutput so we look for those outputs in the inputs table,
+   /// however the block index in the inputs table is the one representing the input not the output we are trying to restore so we have to look it up in the outputs table.
+   /// </summary>
+   private static Task MergeRewindInputsToUnspentTransactionsAsync(MongoData storage, long blockIndex)
+   {
+      const string output = "Output";
+
+      return storage.InputTable.Aggregate()
+         .Match(_ => _.BlockIndex.Equals(blockIndex))
+         .Lookup(storage.OutputTable.CollectionNamespace.CollectionName,
+            new StringFieldDefinition<InputTable>(nameof(Outpoint)),
+            new StringFieldDefinition<OutputTable>(nameof(Outpoint)),
+            new StringFieldDefinition<BsonDocument>(output))
+         .Unwind(_ => _[output])
+         .Project(_ => new
+         {
+            //We need the block index that the output was created on, the rest of the data is the same as input
+            Value = _[nameof(InputTable.Value)],
+            Address = _[nameof(InputTable.Address)],
+            BlockIndex = _[output][nameof(OutputTable.BlockIndex)],
+            Outpoint = new
+            {
+               OutputIndex = _[nameof(Outpoint)][nameof(Outpoint.OutputIndex)],
+               TransactionId = _[nameof(Outpoint)][nameof(Outpoint.OutputIndex)],
```

Comment:
should this be?
`TransactionId = _[nameof(Outpoint)][nameof(Outpoint.TransactionId )]`

Reply:
Yes, I'd rather use the class name than magic strings in the call to mongo